upf/effect.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module defines the Effect class.
+A basic Effect has a fluent and an expression.
+A condition can be added to make it a conditional effect.
+"""
+
+import upf
+import upf.types
+from upf.environment import get_env, Environment
+from upf.fnode import FNode
+from upf.exceptions import UPFTypeError
+from upf.expression import BoolExpression
+from collections import OrderedDict
+from typing import List, Union, Tuple, Optional
+
+KINDS = list(range(0, 3))
+
+(
+    ASSIGN, INCREASE, DECREASE
+) = KINDS
+
+class Effect:
+    def __init__(self, fluent: FNode, value: FNode, condition: FNode, kind: int = ASSIGN):
```

Comment:
Maybe add a default value for `condition`

Reply:
Actually here, to add a default value we would need the environment. Since an Effect is always created in an Action, the Action already auto-promotes to FNode, but giving a default-FNode True value would mean to import the environment also in the Effect class to use the TRUE FNode of the ExpressionManager.

Now, this is not a problem, it can be easily imported, it's just a design choice. Let me know which option you like better.

(This matters only if the user creates an Effect without "using" the class Action, but the action doesn't have a method `
add_effect(Effect)
` so this doesn't seem quite useful to do.)

Reply:
Ah, right! OK, let's leave it like this.